exercise_060_cluster_base/src/main/scala/akka/cluster/pi/LedStripVisualiser.scala

```diff
@@ -0,0 +1,132 @@
+package akka.cluster.pi
+
+/**
+ * Copyright © 2016-2019 Lightbend, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * NO COMMERCIAL SUPPORT OR ANY OTHER FORM OF SUPPORT IS OFFERED ON
+ * THIS SOFTWARE BY LIGHTBEND, Inc.
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * --------------------------------------------------------------------------------
+ * The LedStripController visualises the state of cluster nodes in an Akka Cluster
+ * on an LED strip with at least 8 LEDs.
+ *
+ * It will reflect the state of the cluster as seen be a node on which it runs and
+ * based on events it receives from a `ClusterStatusTracker`.
+ *
+ * Events that contain node-specific information (such as a node moving to the Up
+ * state, use a `NodeLedId` (type Int`), which is the relative node number (or
+ * logical node Id). (e.g. `node-0` -> `0`, ... `node-4` -> `4`)
+ *
+ * As there may be different types of LED strips, this code maps the `NodeLedId`
+ * to a physical LED on an LED strip. This mapping is done via the
+ * `logicalToPhysicalLEDMapping` function.
+ *
+ */
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import org.neopixel._
+
+object LedStripVisualiser {
+
+  def apply(settings: Settings): Behavior[ClusterStatusTracker.NodeState] =
+    Behaviors.setup { context =>
+        new LedStripVisualiser(context, settings)
+          .running(heartbeatLedOn = false, weaklyUpIndicatorOn = false, weaklyUpMembers = Set.empty[Int])
+  }
+}
+
+class LedStripVisualiser(context: ActorContext[ClusterStatusTracker.NodeState],
+                         settings: Settings) {
+
+  import settings._
+
+  private val LeaderLedNumber = 5
+  private val SingletonLedNumber = 6
+
+  private val ledStripDriver = context.spawn(LedStripDriver(settings), "led-strip-driver")
+
+  def running(heartbeatLedOn: Boolean, weaklyUpIndicatorOn: Boolean, weaklyUpMembers: Set[Int]): Behavior[ClusterStatusTracker.NodeState] = Behaviors
```

Comment:
> Right now the indirection doesn't seem to make so much sense yet since the ledstip is the only thing that currently subscribes to the visualizer, but I guess you have something in mind here :D

Indeed I have ;-)

Specifically: as the Raspberry Pi nodes are connected to each other via a physical (UTP) network _but_ are also equipped with a Wifi, we can envision setting up a second (Wifi) network that makes the nodes reachable regardless of the state of connectivity of the nodes via the physical network. You can them imagine having a second visualiser running on each node that can expose the state of the cluster and that can be used by an application that displays that state on a web page. Really cool for demo's, training or talks where it's difficult for all attendees to clearly see the LED-strips.

Reply:
> You keep passing around this state, but it doesn't seem to be used anywhere, is that right?

Nice catch! Thanks!

I forgot to clean this up when I split the driving of the LEDs to the `LedStripDriver` actor

Corrected in the mean time (and pushed).